FlowCrypt/src/main/java/com/flowcrypt/email/ui/activity/fragment/MessageDetailsFragment.kt

```diff
@@ -462,10 +467,49 @@ class MessageDetailsFragment : BaseFragment(), ProgressBehaviour, View.OnClickLi
 
   private fun updateMsgBody() {
     if (msgInfo != null) {
+      updatePgpBadges()
       updateMsgView()
     }
   }
 
+  private fun updatePgpBadges() {
+    msgInfo?.verificationResult?.let { verificationResult ->
+      val badges = mutableListOf<PgpBadgeListAdapter.PgpBadge>()
+
+      if (msgInfo?.encryptionType == MessageEncryptionType.ENCRYPTED) {
+        badges.add(PgpBadgeListAdapter.PgpBadge(PgpBadgeListAdapter.PgpBadge.Type.ENCRYPTED))
+      } else {
+        badges.add(PgpBadgeListAdapter.PgpBadge(PgpBadgeListAdapter.PgpBadge.Type.NOT_ENCRYPTED))
+      }
+
+      if (verificationResult.isSigned) {
+        val badge = when {
+          verificationResult.hasBadSignatures -> {
+            PgpBadgeListAdapter.PgpBadge(PgpBadgeListAdapter.PgpBadge.Type.BAD_SIGNATURE)
+          }
+
+          verificationResult.hasUnverifiedSignatures -> {
+            PgpBadgeListAdapter.PgpBadge(PgpBadgeListAdapter.PgpBadge.Type.CAN_NOT_VERIFY_SIGNATURE)
+          }
+
+          verificationResult.isPartialSigned -> {
+            PgpBadgeListAdapter.PgpBadge(PgpBadgeListAdapter.PgpBadge.Type.ONLY_PARTIALLY_SIGNED)
+          }
+
+          !verificationResult.isPartialSigned && verificationResult.hasMixedSignatures -> {
+            PgpBadgeListAdapter.PgpBadge(PgpBadgeListAdapter.PgpBadge.Type.MIXED_SIGNED)
+          }
```

Comment:
is `!verificationResult.isPartialSigned` here necessary? I thought `when` will match first truthy branch.

Reply:
it depends on our needs. Can we have the following? (three badges)

![image](https://user-images.githubusercontent.com/2863246/146152060-abbc7e07-3942-4d27-8fbc-e2d95a4948c7.png)

Reply:
Not needed. In this situation, show just partially signed which is more important